### ISSUE TITLE
concurrent_reindex regression test still disabled; advisory-lock detach's production behavior unconfirmed

### ISSUE THREAD

**Issue by tier940** [yorishiro-ai/yorishiro#314]
## Background

#283 ("Concurrent reindex_embeddings runs against the same workspace are not serialized") was closed on 2026-08-31. Its fix is `acquire_workspace_reindex_lock` (`src/db.rs:262`), a session-scoped `pg_advisory_lock` held on a connection deliberately detached from the pool (`conn.detach()`), released by dropping the guard. `tests/models/search.rs::concurrent_reindex_runs_serialize_and_consistent_after_lock` is #283's regression test.

## What happened in #313

PR #313 (merged 2026-09-02, `25a78695`) was fixing an unrelated bug (`std::process::exit(0)` killing the whole test binary instead of skipping one test). While chasing CI failures on that PR, the concurrent-reindex test kept failing with:

```
Drop database: Database(PgDatabaseError { ... message: "database \"...\" is being accessed by other users" ... })
```

This is loco_rs's own `DROP DATABASE` teardown (`testing/db.rs:152/156`) hitting a lingering PostgreSQL session. After trying several workarounds (raising `DB_MAX_CONNECTIONS` to 100/200/500, reducing test parallelism, a `pg_terminate_backend` cleanup that was tried then reverted), the PR landed on marking the test `#[ignore] = "flaky in CI: detached connections from advisory lock exhaust pool"` and merged with `acquire_workspace_reindex_lock` itself untouched.

## This did not fix CI

The first `develop` run after #313 merged (run [33585749233](https://github.com/yorishiro-ai/yorishiro/actions/runs/33585749233), job 100109409317) failed with the **exact same symptom** — `db cleanup thread panicked`, `DROP DATABASE ... is being accessed by other users` — but this time on `models::search::sync_embedding_refuses_a_vector_from_a_different_model_than_the_workspace_stamp`, a test that has nothing to do with `acquire_workspace_reindex_lock` or the reindex lock mechanism at all.

That means ignoring the one test did not remove the underlying problem. Either:
- the leak/timing gap in `acquire_workspace_reindex_lock`'s detach-then-drop pattern isn't the sole cause (something broader is leaking connections or exhausting the pool across the suite), or
- the `#[ignore]` comment's diagnosis ("detached connections from advisory lock exhaust pool") was itself incomplete.

## What needs investigating

1. Whether `acquire_workspace_reindex_lock`'s detached connection reliably and promptly closes its PostgreSQL-side session on drop, or whether this can leak/lag under load (asked for in review on #313, not answered before merge).
2. Whether the newly-failing test's failure is the same root cause wearing a different name, or a second, independent issue.
3. Re-enabling `concurrent_reindex_runs_serialize_and_consistent_after_lock` once the actual cause is fixed, since right now #283's own regression coverage is off.

## Related

- #283 (closed, the original issue this test guards)
- #313 (the PR that disabled the test without fixing the cause)

**Comment by tier940** [comment]
## Update: the broader failure is fixed, the original narrow issue is not

#315 (merged 2026-09-02, `ab47728c`) found and fixed a real, separate bug: `spawn_startup_reindex` (`app.rs`) spawned a detached task holding a `ctx.db` connection with no way to signal shutdown or await completion, so `close_app_pools` could close pools while that task was still mid-flight. That's what made the failure land on an unrelated test (`sync_embedding_refuses_a_vector_from_a_different_model_than_the_workspace_stamp`), not the advisory lock at all.

The fix stores the task's `JoinHandle` in `StartupReindexHandle` and `close_app_pools` now calls `shutdown_and_wait()`, which signals and actually awaits the task before closing pools. Verified independently: develop's post-merge CI run ([33594420715](https://github.com/yorishiro-ai/yorishiro/actions/runs/33594420715)) completed clean, `tests/mod.rs`'s binary printed a full `test result: ok. 173 passed; 0 failed; 6 ignored` with no panics, not truncated.

**What #315 explicitly did not touch**, in its own words: "The advisory lock detached connections (in `db.rs::acquire_workspace_reindex_lock`) are a separate leak source that exercises during reindex tests. This fix does not address that."

So two things from the original report are still open:

1. `concurrent_reindex_runs_serialize_and_consistent_after_lock` (#283's regression test) is still `#[ignore]`d. #283's own race-safety claim has no active CI coverage right now.
2. The question asked during #313's review was never answered: does `acquire_workspace_reindex_lock`'s `conn.detach()` reliably and promptly close its PostgreSQL-side session, or can a second reindex call on the same workspace wait longer than expected (or worse) because an idle backend process hasn't yet noticed the detached connection's socket close? This is live production code (`src/workers/reindex.rs`, `src/tasks/reindex_embeddings.rs`), not test-only.

Retitled to reflect this narrower remaining scope.

**Comment by tier940** [comment]
## Recurred on PR #317 (post-#315-merge), same symptom, third distinct test

PR #317's CI failed on this exact symptom twice in a row, after #315 (the `spawn_startup_reindex` leak fix) had already merged and independently verified clean:

- Run [33650927136](https://github.com/yorishiro-ai/yorishiro/actions/runs/33650927136) (2026-09-02 15:48): `models::entities::drift_and_migration_dry_run_see_a_second_version` panics with `Connection pool timed out`, then `db cleanup thread panicked: DROP DATABASE ... is being accessed by other users (There are 4 other sessions using the database)`.
- Run [33654618361](https://github.com/yorishiro-ai/yorishiro/actions/runs/33654618361) (2026-09-02 16:23): same symptom, this time on `models::search::sync_embedding_refuses_a_vector_from_a_different_model_than_the_workspace_stamp` (the exact test named in this issue's original report).

Both runs report **exactly 4 other sessions** holding the throwaway database open at teardown time. That specific number recurring across three separate incidents (the original report, and these two) suggests a fixed number of leaked/lingering connections rather than generic pool pressure from parallel test load.

`concurrent_reindex_runs_serialize_and_consistent_after_lock` is still `#[ignore]`d in both runs, so it is not itself running, but the underlying mechanism it exercises (`acquire_workspace_reindex_lock`'s detached advisory-lock connection) is still live production code reachable from other tests that trigger a reindex, and this issue's still-open question (whether that detach reliably and promptly closes its PostgreSQL-side session) has not been answered.

This is not yet proven to be the advisory-lock path specifically: the two newly-failing tests (`entities::drift_and_migration_dry_run_see_a_second_version`, `search::sync_embedding_refuses_a_vector_from_a_different_model_than_the_workspace_stamp`) do not call `acquire_workspace_reindex_lock` or reindex anything themselves, matching the same pattern #315 found (the failure lands on an unrelated test, not the code path that leaks). Worth checking whether another test running earlier in the same binary (possibly one of PR #317's new SQLite/FTS5 tests) opens a connection it does not close.

**Comment by tier940** [comment]
Resolved by commit 5791083 (part of PR #319).

`WorkspaceReindexLockGuard` now holds `Option<sqlx::PgConnection>` with an explicit `async release().await` that closes the connection and releases the advisory lock on both success and error paths.
`Drop` is a panic-unwind backstop only.

This answers the open question from this issue: whether `acquire_workspace_reindex_lock`'s detached connection reliably and promptly closes its PostgreSQL-side session.
It does, because release is now explicit rather than left to the detached connection's own drop timing.

`#[ignore]` was removed from `concurrent_reindex_runs_serialize_and_consistent_after_lock` in the same commit.

Confirmed on develop's latest CI run after PR #319 merged (run 34005517691): the test passed on both backends.

`test models::search::concurrent_reindex_runs_serialize_and_consistent_after_lock ... ok` (sqlite)
`test models::search::concurrent_reindex_runs_serialize_and_consistent_after_lock ... ok` (postgres)
